### ISSUE TITLE
Suggestion Role Management

### ISSUE THREAD

**Issue by lambersond** [jonas747/yagpdb#741]
Should have the ability to:
* Create
* Delete
* Set color
* Manage permissions

**Comment by ent3r** [comment]
So basically everything you can do from the discord interface?

**Comment by lambersond** [comment]
Correct. I'd like to have the ability to automate that process.

**Comment by ent3r** [comment]
What is there to automate? Make a command for automatic role creation? Is that what you want?
What is the use case for automating role creation?

**Comment by lambersond** [comment]
Yes, that is exactly what I want. To be able to spin up & down roles as needed.

The use case goes hand with https://github.com/jonas747/yagpdb/issues/740

I have a server that hosts tabletop RPG groups, and the person running the game, gamemaster, will want to spin up a private channel for them and their players. 

The idea is that when a game master wants to start a new game, they could just execute the following custom command and it'd spin up a new channel for them.

```
{{$args := parseArgs 1 "Campaign Name" (carg "string" "Campaign name")}}
{{$options := (cslice 
"A hooded figure, shrouded in darkness, approaches you..." 
"Above the tavern, a dragon's roar can be heard..." 
"An unmarked letter appears at your feet. It emanates dark smoke and voice is heard in your head..."
"A dwarf stumbles in and with his dying breath whispers, \"the guards have fallen...\""
"Darkness swirls over head. The floorboards shatter as hands reach out from the ground..."
"As a child, your town was ravaged by the green skull banner. 30 years later you see the banner approaching in the distance..."
)}}
{{$rInt := randInt 6}}

{{$campaignName := (title ($args.Get 0))}}
{{$roleName := (joinStr "" "5E " (title ($args.Get 0)) "")}}
{{$channelName := (reReplace " " ($args.Get 0) "-")}}
{{$type := "5E"}} {{/*Add as arg in future */}}

{{/* $channelID := create channel in category => return ID */}}
{{/*$roleID := create role & make mentionable => return ID */}}

{{/*DB format: <.User.ID>_<type[string]>_<name[string]>_<lfg[bool]>_<archived[bool]>_<$channel_ID>_<$role_ID> */}}
{{$entry:= (joinStr "" (.User.ID) "_" $type "_" $campaignName "_true_false_unknown_unknown" "")}}

{{/* change from $channelName to $channelID when able */}}
{{dbSet 1001 $channelName $entry}}

{{ $embed := cembed
"description" (joinStr "" "** \n" $campaignName "**\n\n**Game Master: **" (.User.Mention) "\n**Ruleset: **" $type "\n**Status: ** [OPEN]" "" )
"color" 9021952
"timestamp" currentTime
"author" (sdict "name" (index $options $rInt) "icon_url" "https://cdn.discordapp.com/emojis/725168618257514506.png?v=1" )
}}

{{sendMessageNoEscape nil (complexMessage "content" (joinStr "" (mentionRoleID 301495699789578242 ) " a new campaign just opened!" "" ) "embed" $embed)}}
```

The command doesn't have a purpose currently because I would still need to create the role & channel for the member manually.

Once the game was completed.. months down the road, the GM could just run `remove-campaign <x>` and it would removed the channel & role.

**Comment by CaioStoduto** [comment]
I think it's really helpful, and I'd like to see an option to set the role position too.

**Comment by Longhouse81** [comment]
I think you already have kinda an option to do this: 
you can role a member and create a ticket, create custom CC to add users to the ticket that is able to be executed by that role... 
At this point, you would be limited to 10 simultaneous channels using the tickets. Should work not?

**Comment by CaioStoduto** [comment]
That's a really good solution to his problem. However, the ability to create a role, modify the role (change colour, position, etc.), using CC would be really useful.

**Comment by Longhouse81** [comment]
Let me comment on that:
Color roles: i never see the purpose why people use permissions in combination with color on a role in general, while you can create different color roles and separate permissions from them you don't have the need for any position or anything, no perms on a color role just means a color, making it they can be set above all regular roles and you can change them easy...
Creating and modify roles: While every permission system on operating systems works with stacking permissions using the AGDLP method I don't see why you would have the need to combine all perms into 1 role, you create multiple roles with permissions that can be added, then the bot can fully automate that again you instead of giving 1 role you add 3 or 5 depending what rights you want to give to the user, it more clear that way.
Example: your mod can ban and kick, a super moderator need to have access to bot, you add only manage server to super moderator and give the super mod also the moderator role. he can do now what is needed to perform his tasks without you having the need to copy the moderator role and adding manage server to it, there is no need to adjust all other parameters and it easy to manage because you know the super mod can only manage server more then a regular mod...
It would be nice to be able to give/change permissions, coding wise for the bot it would cost many hours in building in wrong use by code, making it a lot of work while in a way you are already able to perform actions like this by adding a permission role to a user using the AGDLP method

**Comment by lambersond** [comment]
I don't think I've clearly stated the issue. So let me try again. I want to leverage YAGBDB to create new roles so that I can assign the role to a [new channel](https://github.com/jonas747/yagpdb/issues/740) and a small group of people to give them a private channel with a "channel leader" for back of better words.

I get creating roles and channels can be done through the interface, but so can everything else that the bot does. The purpose of this ticket is to add the ability to spin up and down roles as needed so a person doesn't have to manually do it. 

I would agree that creating a ticket and linking the users to the ticket would work if I just wanted to link small group together, but that's not the purpose of this request.

**Comment by Longhouse81** [comment]
i completely understand what you say, tho the limit of tickets is 10 you can actually do this yourself, create x amount of channels you want to have available, create for each of those channels 1 roles a "member" and a general role for the whole list that can trigger the commands the "team leader", make a DB entry that tracks available channels for leaders and members to use, you have the option to add people, same as the ticket, but no limitation of 10 channels... 
While you ask from the devs to build in a check you not go over the channel limits, roles limits, creation of channels limit, all discord limitation stuff they need to do, what ends up with the same result, only your CC take less time because you only have to work with bot limitations and not discord limitations, so creating what you want with CC would take 1% of the time they need to spend on creating stuff you can already do, what in my opinion is wasted time...
Then we come to the part where I ask you if you ever have seen the amount of code needed in a CC just to read the permissions, that would leave no space for actual other code when you adjust permissions. Coming down to the part you also ask for functions from the dev where it all comes down to the part: you spent more time and the dev spent more time then needed while you already can do what you want...
-the roles you want to be created can be premade by you, copied x amount of times depending on how many channels you want, that cost you like 10 minutes while coding it would be 10x that time
-the channels is the same, rename them on start and you good to go
-when you use the ticket system you don't have to deal with all that stuff, even have more functionality because you can call 5 ticket functions in 1 CC, then you loop back to the same CC if you need more (adding users would be possible to add 10x 6 users per minute that way)
- then the documentation would take even more time...
-"this ticket is to add the ability to spin up and down roles as needed so a person doesn't have to manually do it." still comes down the fact Discord is limited to 250 roles and 500 channels/ 50 channels per category...
While the ticket system add users to a channel and not roles you can set you CC to ignore all other channels in your discord and have an easy only 1 role needed for all channels with a channel leader making it you would be able to save on 10 tickets 10 roles used...

My conclusion is basically: 
You maybe want to spend more time on creating scripts instead of creating and copying some roles (copy roles work easy, you set all perms on the everyone role and when you create a role you get those perms on the new role), then you can copy some channels depending the amount you want and place them in a category that you easy collapse, so when no activity in those they not visible. And you don't use the build-in ticket feature when you stay under 10 channels needed...
You seem not to understand if you want the bot to create roles they need to be reserved in the total of 250 roles, making it you actually better just create them instead of reserving them...
I think it's unfair to ask at the devs to do 100x the amount of coding because you want to spent more time and not use the existing features that the bot has...

**Comment by lambersond** [comment]
I'm well aware of the role and channel limitations on a discord server. I'm also not asking to handle error checking. I'm also well aware that creating a role, channel, or message is a POST request to discords API. That can be passed through, similar to any other error. I can spin up generic channels and roles and and assign them to each other and use tickets and the db functionality to handle who can see those channels. That is easy to script out and would be an interim solution. However, that's like using a saw to hammer in a nail. Just because you can do it doesn't mean it's the proper tool for the job.

**Comment by Longhouse81** [comment]
Euh if you don't error handle your request to an API the API times you out in x amount of cases. 1 user can time out the bot then and that need to be avoided at all cost...

**Comment by lambersond** [comment]
okay.. so listroles gives a count so the methodology exists to get count, is `if (roles.count le max) { \\ send request } else {ignore}` you could even have an internal cooldown at server level to reduce frequency of requests and chance of ddos attacks.
